File: src/questions/makeFactorNumberQuestion.ts
import { getRandomInt } from "../util/getRandomInt"

type FactorNumberQuestionOpts = {
  min?: number
  max?: number
}

type FactorNumberQuestion = {
  questionNumber: number
  products: number[][]
}

export const makeFactorNumberQuestion = (
  opts: FactorNumberQuestionOpts = {},
): FactorNumberQuestion => {
  const { min = 1, max = 1000 } = opts
  const questionNumber = getRandomInt({ max, min })
  const question: FactorNumberQuestion = {
    questionNumber,
    products: [],
  }

  for (let i = 2; i * i <= questionNumber; i++) {
    // for (let i = questionNumber - 1; i * i >= questionNumber; i--) {
    if (questionNumber % i === 0) {
      question.products.push([i, questionNumber / i])
    }
  }

  return question
}

File: cpp/src/questions/makeFactorNumberQuestion.cpp
#include "questions/makeFactorNumberQuestion.h"
#include "util/getRandomInt.h"

FactorNumberQuestion makeFactorNumberQuestion(FactorNumberQuestionOpts opts)
{
  const int questionNumber = getRandomInt({
      .min = opts.min,
      .max = opts.max,
  });
  FactorNumberQuestion question{
      .questionNumber = questionNumber,
      .products = {},
  };

  for (int i = 2; i * i <= questionNumber; i++)
  {
    if (questionNumber % i == 0)
    {
      int secondProduct = questionNumber / i;
      question.products.push_back({i, secondProduct});
    }
  }

  return question;
}
